src/rat/plugins/forecasting.py

```diff
@@ -341,6 +339,373 @@ def get_GFS_data(basedate, lead_time, basin_bounds, gfs_dir):
+            forecast_outflow = forecast_outflow.merge(forecast_outflow_cust_wl)
+
+    ######### If dels_scenario = RC (Rule Curve Scenario ############ 
+    # S/Smax ratio for base date and base date + 15 days is computed.
+    # delS/Smax is obtained and multiplied with known Smax to get permissible delS.
+    # if delS is +ve, inflow is accumulated. If delS is -ve, outflow is produced by evenly releasing required inflow over the 15 day period.
+    if(dels_scenario is None or dels_scenario == 'RC'):
+        if rule_curve is not None:
+            forecast_outflow_rule_curve = forecast_scenario_rule_curve(forecast_outflow, initial_sa, base_date, base_date_15lead, rule_curve, s_max, aec_data)
+            forecast_outflow = forecast_outflow.merge(forecast_outflow_rule_curve)
+
+    ######### If dels_scenario = ST (Storage Scenario ############ 
+    # Storage change scenarios take a list of %Smax values as permissible change in storage. Eg. [1,2.5,5] (1%, 2.5%, and 5% of Smax)
+    # Outflows, delH, delS, sarea are generated for all %Smax values as per similar logic as RC scenario.
+    # Custom storage value in MCM may also be provided. (release: -ve or store: +ve)
+    if(dels_scenario is None or dels_scenario == 'ST'):
+        forecast_outflow_st = forecast_scenario_st(forecast_outflow, initial_sa, cust_st, s_max, aec_data, st_percSmax)
+        forecast_outflow = forecast_outflow.merge(forecast_outflow_st)
+
+    if(output_path is not None):
+        forecast_outflow.to_pandas().to_csv(output_path)
+
+    return forecast_outflow
+
+def forecast_outflow(
+    basedate, lead_time, basin_data_dir, reservoir_shpfile, reservoir_shpfile_column_dict,
+    rule_curve_dir,
+    scenarios = ['GC', 'GO', 'RC', 'ST'],
+    st_percSmaxes = [0.5, 1, 2.5],
+):
+    reservoirs = gpd.read_file(reservoir_shpfile)
+    reservoirs['Inflow_filename'] = reservoirs[reservoir_shpfile_column_dict['unique_identifier']].astype(str)
+
+    for res_name in reservoirs['Inflow_filename'].tolist():
+        res_scenarios = scenarios.copy()
+        forecast_inflow_fp = basin_data_dir / 'rat_outputs' / 'forecast_inflow' / f'{basedate:%Y%m%d}' / f'{res_name}.csv'
+        forecast_evap_fp = basin_data_dir / 'rat_outputs' / 'forecast_evaporation' / f'{basedate:%Y%m%d}' / f'{res_name}.csv'
+        sarea_fp = basin_data_dir / 'gee' / 'gee_sarea_tmsos' / f'{res_name}.csv'
+        aec_fp = basin_data_dir / 'final_outputs' / 'aec' / f'{res_name}.csv'
+        output_fp = basin_data_dir / 'rat_outputs' / 'forecast_outflow' / f'{basedate:%Y%m%d}' / f'{res_name}.csv'
+        output_fp.parent.mkdir(parents=True, exist_ok=True)
+
+        s_max = reservoirs[reservoirs[reservoir_shpfile_column_dict['unique_identifier']] == res_name]['CAP_MCM'].values[0]
+        grand_id = reservoirs[reservoirs[reservoir_shpfile_column_dict['unique_identifier']] == res_name]['GRAND_ID'].values[0]
```

Comment:
It seems to be specific for GRAND. As RAT is more generic, can we take the column name of capacity from config? 'ID' column name is already extracted from config.

Reply:
Thanks for the suggestion. The reservoir ID and capacity of the reservoir will now be picked from the reservoir shapefile based on the `column_id` and `column_capacity` options specified in the config file.